### PR TITLE
Do not bind webhook port if webhooks are not present

### LINKED ISSUE #120043: cloud-provider binds ports 10258+10260 even if no webhooks are configured
### What happened?

Starting in v1.27.0, programs using `kubernetes/cloud-provider` started binding ports 10258 & 10260 (TCP) even if the associated functionality (webhooks) are not configured and not used. This fails if the cloud-controller-manager is run using `hostNetwork: true` and the ports are already in use (rolling deployment):

```
failed to create listener: failed to listen on 0.0.0.0:10260: listen tcp 0.0.0.0:10260: bind: address already in use
```

### What did you expect to happen?

I did not expect that ports are opened for a feature behind a disabled feature flag.

### How can we reproduce it (as minimally and precisely as possible)?

Bind port `10260` on the host:
```
nc -l 10260
```

Run any cloud-controller-manager locally with the required flags set, this example uses `hcloud-cloud-controller-manager`. The ccm must use k/cloud-provider v1.27.x:

```
git clone git@github.com:hetznercloud/hcloud-cloud-controller-manager.git
cd hcloud-cloud-controller-manager
git switch --detach v1.17.1
go get
go run main.go --cloud-provider=hcloud --kubeconfig ../../hcloud-cloud-controller-manager/hack/.kubeconfig-dev
```

It fails with:

```
I0818 12:04:31.870440  133789 serving.go:348] Generated self-signed cert in-memory
I0818 12:04:32.084205  133789 serving.go:348] Generated self-signed cert in-memory
failed to create listener: failed to listen on 0.0.0.0:10260: listen tcp 0.0.0.0:10260: bind: address already in use
Error: failed to create listener: failed to listen on 0.0.0.0:10260: listen tcp 0.0.0.0:10260: bind: address already in use
[help output]
```

#### Traces

Stack trace for port `10260`:

```go
options.createListener (webhook.go:193) k8s.io/cloud-provider/options
options.(*WebhookServingOptions).ApplyTo (webhook.go:169) k8s.io/cloud-provider/options
options.(*CloudControllerManagerOptions).ApplyTo (options.go:202) k8s.io/cloud-provider/options
options.(*CloudControllerManagerOptions).Config (options.go:301) k8s.io/cloud-provider/options
app.NewCloudControllerManagerCommand.func1 (controllermanager.go:99) k8s.io/cloud-provider/app
cobra.(*Command).execute (command.go:916) github.com/spf13/cobra
cobra.(*Command).ExecuteC (command.go:1040) github.com/spf13/cobra
cobra.(*Command).Execute (command.go:968) github.com/spf13/cobra
main.main (main.go:57) main
runtime.main (proc.go:250) runtime
runtime.goexit (asm_amd64.s:1598) runtime
 - Async Stack Trace
<autogenerated>:2
```

Stack trace for port `10258`:

```go
options.CreateListener (serving.go:360) k8s.io/apiserver/pkg/server/options
options.(*SecureServingOptions).ApplyTo (serving.go:244) k8s.io/apiserver/pkg/server/options
options.(*SecureServingOptionsWithLoopback).ApplyTo (serving_with_loopback.go:44) k8s.io/apiserver/pkg/server/options
options.(*CloudControllerManagerOptions).ApplyTo (options.go:206) k8s.io/cloud-provider/options
options.(*CloudControllerManagerOptions).Config (options.go:301) k8s.io/cloud-provider/options
app.NewCloudControllerManagerCommand.func1 (controllermanager.go:99) k8s.io/cloud-provider/app
cobra.(*Command).execute (command.go:916) github.com/spf13/cobra
cobra.(*Command).ExecuteC (command.go:1040) github.com/spf13/cobra
cobra.(*Command).Execute (command.go:968) github.com/spf13/cobra
main.main (main.go:57) main
runtime.main (proc.go:250) runtime
runtime.goexit (asm_amd64.s:1598) runtime
 - Async Stack Trace
<autogenerated>:2
```

#### Workaround

The port binding can be disabled by setting the flags `--webhook-secure-port 0` && `--secure-port 0`, but setting both at the same time causes a validation error:

```
Error: --webhook-secure-port cannot be the same value as --secure-port
```

#### References

The associated changes are from https://github.com/kubernetes/kubernetes/pull/108838  
Originally noticed and debugged in https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/934

### Kubernetes version

<details>

Not relevant, as the issue happens without kubectl and without a running Kubernetes cluster.

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4", GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"archive", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

</details>

### Cloud provider

<details>
Hetzner Cloud

Debugged using hcloud-cloud-controller-manager main & 1.17.1 using k8s.io/cloud-provider v0.27.4
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
BUILD_ID="23.11.20230816.caac0eb"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 23.11 (Tapir)"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="23.11 (Tapir)"
VERSION_CODENAME=tapir
VERSION_ID="23.11"
$ uname -a
Linux nephele 6.1.45 #1-NixOS SMP PREEMPT_DYNAMIC Fri Aug 11 10:08:27 UTC 2023 x86_64 GNU/Linux
```
</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by kmala**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Don't bind to the webhook port if there are no webhooks. Since we don't start the server if there are no webhooks, there is no need to bind.

#### Which issue(s) this PR fixes:

Fixes #120043 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
/ok-to-test

This seems to come from https://github.com/kubernetes/kubernetes/pull/108838

/assign @nckturner @cheftako 

that may have more context based on `git blame`

File: staging/src/k8s.io/cloud-provider/options/options_test.go [hunk 1]
```diff
@@ -448,6 +449,135 @@ func TestCreateConfig(t *testing.T) {
...
+		"--leader-elect-retry-period=5s",
+		"--master=192.168.4.20",
+		"--min-resync-period=100m",
+		"--node-status-update-frequency=10m",
+		"--profiling=false",
+		"--route-reconciliation-period=30s",
+		"--use-service-account-credentials=false",
+	}
+	err = fs.Parse(args)
+	assert.Nil(t, err, "unexpected error: %s", err)
```
Comment (elmiko) [inline staging/src/k8s.io/cloud-provider/options/options_test.go:501]: looks like the linter wants `assert.NoError` here

File: staging/src/k8s.io/cloud-provider/options/options_test.go [hunk 2]
```diff
@@ -448,6 +449,135 @@ func TestCreateConfig(t *testing.T) {
...
+
+	fs.VisitAll(func(f *pflag.Flag) {
+		fmt.Printf("%s: %s\n", f.Name, f.Value)
+	})
+
+	c, err := s.Config([]string{"foo", "bar"}, []string{}, nil, []string{"foo", "bar", "baz"}, []string{})
+	assert.Nil(t, err, "unexpected error: %s", err)
```
Comment (elmiko) [inline staging/src/k8s.io/cloud-provider/options/options_test.go:508]: same `assert.NoError`

File: staging/src/k8s.io/cloud-provider/options/options_test.go [hunk 3]
```diff
@@ -448,6 +449,135 @@ func TestCreateConfig(t *testing.T) {
...
+
+	for _, f := range s.Flags([]string{""}, []string{""}, nil, []string{""}, []string{""}).FlagSets {
+		fs.AddFlagSet(f)
+	}
+
+	tmpdir, err := os.MkdirTemp("", "options_test")
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
+	defer os.RemoveAll(tmpdir)
```
Comment (elmiko) [inline staging/src/k8s.io/cloud-provider/options/options_test.go:468]: linter complaining about the return error return value on `RemoveAll`

File: staging/src/k8s.io/cloud-provider/options/webhook.go [hunk 4]
```diff
@@ -151,14 +151,18 @@ func (o *WebhookServingOptions) Validate() []error {
...
-func (o *WebhookServingOptions) ApplyTo(cfg **server.SecureServingInfo) error {
+func (o *WebhookServingOptions) ApplyTo(cfg **server.SecureServingInfo, webhookCfg config.WebhookConfiguration) error {
 	if o == nil {
 		return nil
 	}
 
 	if o.BindPort <= 0 {
 		return nil
 	}
+	//no need to bind to the address if there are no webhook enabled.
```
Comment (elmiko) [inline staging/src/k8s.io/cloud-provider/options/webhook.go:162]: linter wants a space after `//`

**Review by elmiko — COMMENTED** [review]
this solution makes sense to me, just a few linter cleanups needed. (i wasn't ever aware of `assert.NoError`, TIL!)

**Comment by elmiko** [comment]
thanks @kmala !

/lgtm

cc @cheftako for approval

**Pull request merged**